src/nullable.ts

```diff
@@ -0,0 +1,35 @@
+import { ModelValueType } from './glossary'
+import { Relation } from './relations/Relation'
+
+export type NullableGetter<ValueType extends ModelValueType> =
+  () => ValueType | null
+
+export class NullableProperty<ValueType extends ModelValueType> {
+  public getValue: NullableGetter<ValueType>
+
+  constructor(getter: NullableGetter<ValueType>) {
+    this.getValue = getter
+  }
+}
+
+export class NullableRelation<RelationType extends Relation<any, any, any>> {
+  public relation: RelationType
+
+  constructor(relation: RelationType) {
+    this.relation = relation
+  }
+}
+
+export function nullable<ValueType extends ModelValueType>(
+  value: NullableGetter<ValueType>,
+): NullableProperty<ValueType>
+export function nullable<RelationType extends Relation<any, any, any>>(
+  value: RelationType,
+): NullableRelation<RelationType>
+export function nullable(value: any) {
```

Comment:
I believe you don't have to annotate the `value` here, it should be inferred from the overloads above. Isn't that the case?

Reply:
It isn't inferred unfortunately but I did manage to improve the type by using a union of the different value types. I feel like typescript should've handled that for me like you suggest haha. Ah well 🤷‍♂️